cmake/test/gtest.cmake

```diff
@@ -85,97 +85,261 @@ function(catkin_add_executable_with_gtest target)
+  find_file(_GMOCK_SOURCES "gmock.cc"
+            PATHS ${src_paths}
+            NO_DEFAULT_PATH NO_CMAKE_FIND_ROOT_PATH
+  )
+
+  # If we found gmock, ensure gtest is contained within it
+  if(_GMOCK_INCLUDES AND _GMOCK_SOURCES)
+    get_filename_component(SOURCE_DIR ${_GMOCK_SOURCES} PATH)
+    get_filename_component(BASE_DIR ${SOURCE_DIR} PATH)
+
+    get_filename_component(INCLUDE_DIR ${_GMOCK_INCLUDES} PATH)
+    get_filename_component(INCLUDE_DIR ${INCLUDE_DIR} PATH)
+
+    set(${found} TRUE PARENT_SCOPE)
+    set(${base_dir} ${BASE_DIR} PARENT_SCOPE)
+    set(${include_dir} ${gtest_include_dir} ${INCLUDE_DIR} PARENT_SCOPE)
+    set(${lib_dir} ${CMAKE_BINARY_DIR}/gmock PARENT_SCOPE)
+    set(${libs} "gmock" PARENT_SCOPE)
+    set(${main_libs} "gmock_main" PARENT_SCOPE)
+  endif()
+endfunction()
+
+find_package(GMock QUIET)
+if(NOT GMOCK_FOUND)
+  find_package(GTest QUIET)
+  if(NOT GTEST_FOUND)
+    # only add gmock/gtest directory once per workspace
+    if(NOT TARGET gtest AND NOT TARGET gmock)
+      # Fall back to system-installed gmock source (e.g. Ubuntu)
+      set(_include_paths "/usr/include/gmock")
       if(CATKIN_TOPLEVEL)
-        message(STATUS "gtest not found, C++ tests can not be built. Please install the gtest headers globally in your system or checkout gtest (by running 'svn checkout http://googletest.googlecode.com/svn/tags/release-1.6.0 gtest' in the source space '${CMAKE_SOURCE_DIR}' of your workspace) to enable gtests")
-      else()
-        message(STATUS "gtest not found, C++ tests can not be built. Please install the gtest headers globally in your system to enable gtests")
+        # Ensure current workspace is searched before system path
+        list(INSERT _include_paths 0 "${CMAKE_SOURCE_DIR}/gmock/include/gmock")
+      endif()
+
+      set(_source_paths "/usr/src/gmock/src")
+      if(CATKIN_TOPLEVEL)
+        # Ensure current workspace is searched before system path
+        list(INSERT _source_paths 0 "${CMAKE_SOURCE_DIR}/gmock/src")
+      endif()
+
+      catkin_find_gmock_source("${_include_paths}" "${_source_paths}" gmock_found
+                               gmock_base_dir gmock_include_dir gmock_lib_dir
+                               gmock_libs gmock_main_libs)
+
+      # If we found gmock, set it up to be built (which will also build gtest,
+      # since it's bundled)
+      if(gmock_found)
+        set(GMOCK_FROM_SOURCE_FOUND ${gmock_found} CACHE INTERNAL "")
+        set(GMOCK_FROM_SOURCE_INCLUDE_DIRS ${gmock_include_dir} CACHE INTERNAL "")
+        set(GMOCK_FROM_SOURCE_LIBRARY_DIRS ${gmock_lib_dir} CACHE INTERNAL "")
+        set(GMOCK_FROM_SOURCE_LIBRARIES ${gmock_libs} CACHE INTERNAL "")
+        set(GMOCK_FROM_SOURCE_MAIN_LIBRARIES ${gmock_main_libs} CACHE INTERNAL "")
+
+        set(_CATKIN_SKIP_INSTALL_RULES TRUE)
+        function(install)
+          if(_CATKIN_SKIP_INSTALL_RULES)
+            return()
+          endif()
+          _install(${ARGN})
+        endfunction()
```

Comment:
Please keep the comment why this is necessary.

Same below.

Reply:
Done in 4d698414adb6801bcbf0668cba56f122b6109d24